server/models/campaign.py

```diff
@@ -166,6 +167,21 @@ def as_dict(self):
             self, recurse=False, exclude=[Note.room, Note.location, Note.user]
         )
 
+class Marker(BaseModel):
```

Comment:
I see some issues here.  You declare both room and location.  This can only lead to potential bugs (e.g. a marker with room=a and location=l  where l is a location in room b).  You have access to room via the location refkey, so there is no need for a direct link to room.

Coincidentally there is actually also no real reason to have a direct link to location.  What I think you actually want, is to have the uuid be a reference to a shape directly, as a Marker in its current form is always associated with a particular shape.  Take for example ShapeLabel in shape.py, it has 2 direct foreign keys.  One to shape and one to label.  What we want here ultimately is 1 reference to shape and 1 to user, nothing more.

Reply:
I'm so annoyed (with myself) at the moment. I had the process working with strings instead of objects, then when hitting refresh, all markers were read as objects. So I changed the class to return a string instead of objects, now that works, now I can't write to the db. Whenever I want to, the server responds with "Task exception was never retrieved" on executing the `Marker.get_or_none(uuid=data)` where I'm sure data is a string (according to the log), not an object but when I change data to data.uuid, I get "'str' object has no attribute 'uuid'". That makes sense but I have no clue how to change the marker.py contents to read the db as string instead of object or to get rid of the warning.

Reply:
Nevermind...... just a contraint on the table that I completely overlooked. All's well again...